File: binary_tree.cpp
/*
 * Copyleft GTO Inc. All rights reversed.
 */

#include <iostream>

#include <stack>
#include <queue>

using namespace std;

namespace gtlib {

    template<class T>
    class binary_tree {

        struct Node {
            Node(const T &value, Node *left = nullptr, Node *right = nullptr) : value(value), left(left), right(right) {
            }

            const T value;
            Node *left;
            Node *right;
        };

        Node *root_;

    public:
        binary_tree() : root_(nullptr) {
        }

        ~binary_tree() {
            destroy_it(root_);
        }

        Node *add(const T &v, Node *parent = nullptr) {

            if (!parent) {
                parent = new Node(v);
                if (!root_) {
                    root_ = parent;
                }
            } else if (v < parent->value) {
                parent->left = add(v, parent->left);
            } else if (v > parent->value) {
                parent->right = add(v, parent->right);
            }

            return parent;
        }

        Node *add_it(const T &v, Node *parent = nullptr) {

            Node *tmp = parent;
            Node *prev = nullptr;

            while (tmp) {
                prev = tmp;
                if (v < tmp->value) {
                    tmp = tmp->left;
                } else {
                    tmp = tmp->right;
                }
            }

            tmp = new Node(v);

            if (!root_) {
                root_ = tmp;
            } else if (v < prev->value) {
                prev->left = tmp;
            } else if (v > prev->value) {
                prev->right = tmp;
            }
            return tmp;
        }

        bool search(const T &v, Node *parent = nullptr) {
            if (parent == nullptr) {
                return false;
            } else if (v < parent->value) {
                return search(v, parent->left);
            } else if (v > parent->value) {
                return search(v, parent->right);
            } else {
                return true;
            }
        }

        bool search_it(const T &v, Node *parent = nullptr) {

            stack < Node * > tovisit;
            tovisit.push(parent);

            while (tovisit.size() > 0) {

                const Node *curr = tovisit.top();
                tovisit.pop();

                if (curr) {

                    tovisit.push(curr->left);
                    tovisit.push(curr->right);

                    if (curr->value == v) {
                        return true;
                    }
                }

            }

            return false;
        }

        void destroy(Node *parent) {
            if (parent) {
                destroy(parent->left);
                destroy(parent->right);

                delete parent;
            }
        }

        void destroy_it(Node *parent) {

            stack < Node * > tovisit;
            tovisit.push(parent);

            while (tovisit.size() > 0) {

                Node *curr = tovisit.top();
                tovisit.pop();

                if (curr) {

                    tovisit.push(curr->left);
                    tovisit.push(curr->right);

                    delete curr;
                }
            }
        }

        size_t get_height(Node *parent = nullptr) {
            if (!parent) {
                return 0;
            } else {
                return 1 + max(get_height(parent->left), get_height(parent->right));
            }
        }

        size_t get_height_it(Node *parent = nullptr) {

            size_t height = 0;

            queue < Node * > visited;
            visited.push(parent);
            visited.push(nullptr);

            while (Node *f = visited.front()) {
                visited.pop();
                if (f->left) {
                    visited.push(f->left);
                }
                if (f->right) {
                    visited.push(f->right);
                }
                if (!visited.front()) {
                    visited.pop();
                    visited.push(nullptr);
                    height++;
                }
            }

            return height;
        }
    };

}

template<typename node_type, typename output_type>
void print_binary_tree(output_type &out, node_type *n, size_t depth = 0, char connector = '+') {
    if (!n) {
        return;
    }

    print_binary_tree(out, n->right, depth + 1, '/');

    for (int i = depth; i--;) {
        out << "    ";
    }
    out << ' ' << connector << "--" << n->value << endl;

    print_binary_tree(out, n->left, depth + 1, '\\');
}

int main() {

    gtlib::binary_tree<int> tree;

    auto *root = tree.add_it(4);

    tree.add_it(2, root);
    tree.add_it(1, root);
    tree.add_it(3, root);
    tree.add_it(6, root);
    tree.add_it(5, root);
    tree.add_it(7, root);
    print_binary_tree(cout, root);

    tree.add(10, root);
    print_binary_tree(cout, root);

    cout << tree.search_it(2, root) << endl;
    cout << tree.search(7, root) << endl;

    cout << tree.search_it(10, root) << endl;

    cout << "root height: " << tree.get_height(root) << " - " << tree.get_height_it(root) << endl;

    return 0;
}

File: binary_tree.cs
/*
* Copyleft 1979-2013 GTO Inc. All rights reversed.
*/

using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace gtlib
{
    public class BinaryTree<T>
    {
        public class Node
        {
            public T value;
            public Node left;
            public Node right;
        }

        private Node root = null;

        private Node Search(T value, Node node)
        {
            if (node != null)
            {
                Comparer<T> cmp = Comparer<T>.Default;
                if (cmp.Compare(value, node.value) < 0)
                    return Search(value, node.left);
                else if (cmp.Compare(value, node.value) > 0)
                    return Search(value, node.right);
            }
            return node;
        }

        public Node Search(T value)
        {
            return Search(value, root);
        }

        private Node Add(T value, Node node)
        {
            Comparer<T> cmp = Comparer<T>.Default;
            if (node == null)
                node = new Node { value = value };
            else if (cmp.Compare(value, node.value) < 0)
                node.left = Add(value, node.left);
            else if (cmp.Compare(value, node.value) > 0)
                node.right = Add(value, node.right);
            return node;
        }

        public void Add(T value)
        {
            root = Add(value, root);
        }

        private Node Ancestor(Node q, Node r)
        {
            if (r.right != null)
                r.right = Ancestor(q, r.right);
            else
            {
                q.value = r.value;
                r = r.left;
            }
            return r;
        }

        private Node Remove(T value, Node node)
        {
            if (node != null)
            {
                Comparer<T> cmp = Comparer<T>.Default;
                if (cmp.Compare(value, node.value) < 0)
                    node.left = Remove(value, node.left);
                else if (cmp.Compare(value, node.value) > 0)
                    node.right = Remove(value, node.right);
                else
                {
                    if (node.right == null)
                        node = node.left;
                    else if (node.left == null)
                        node = node.right;
                    else
                        node.left = Ancestor(node, node.left);
                }
            }
            return node;
        }

        public void Remove(T value)
        {
            root = Remove(value, root);
        }

        private void Print(StringBuilder sb, Node node, int depth, char connector)
        {
            if (node == null)
                return;

            Print(sb, node.right, depth + 1, '/');

            for (int i = depth; i-- > 0; )
                sb.Append('\t');

            sb.AppendFormat(" {0} -- {1}\n", connector, node.value);

            Print(sb, node.left, depth + 1, '\\');
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();
            Print(sb, root, 0, '+');
            return sb.ToString();
        }
    }
}
